### PR TITLE
split up discovery for readability and stubbability

### PR DESCRIPTION
**Pull request by grosser**
Issue 1:
testing kubernetes is messy because every test needs to stub out calls to the discover endpoint

... I want to be able to easily stub out the discover response in my tests by overriding or stubbing `fetch_entities`

Issue 2:
every new client needs to make a discover call before being useable

... I want to be able to share `@entities` accross all instances or KubeClient by adding some class-level caching ... can do this by overriding `load_entities`

related to https://github.com/abonas/kubeclient/pull/185

@moolitayer @simon3z @enoodle 
/cc @jonmoter @irwaters

### CODE REVIEW

**Comment by grosser** [comment]
FYI: current stub

```Ruby

  before do
    # https://github.com/abonas/kubeclient/pull/200
    discover = {
      "kind"=>"APIResourceList", 
      "groupVersion"=>"v1", 
      "resources"=>[
        {"name"=>"bindings", "namespaced"=>true, "kind"=>"Binding"}, 
        {"name"=>"componentstatuses", "namespaced"=>false, "kind"=>"ComponentStatus"}, 
        {"name"=>"configmaps", "namespaced"=>true, "kind"=>"ConfigMap"}, 
        {"name"=>"endpoints", "namespaced"=>true, "kind"=>"Endpoints"}, 
        {"name"=>"events", "namespaced"=>true, "kind"=>"Event"}, 
        {"name"=>"limitranges", "namespaced"=>true, "kind"=>"LimitRange"}, 
        {"name"=>"namespaces", "namespaced"=>false, "kind"=>"Namespace"}, 
        {"name"=>"namespaces/finalize", "namespaced"=>false, "kind"=>"Namespace"}, 
        {"name"=>"namespaces/status", "namespaced"=>false, "kind"=>"Namespace"}, 
        {"name"=>"nodes", "namespaced"=>false, "kind"=>"Node"}, 
        {"name"=>"nodes/proxy", "namespaced"=>false, "kind"=>"Node"}, 
        {"name"=>"nodes/status", "namespaced"=>false, "kind"=>"Node"}, 
        {"name"=>"persistentvolumeclaims", "namespaced"=>true, "kind"=>"PersistentVolumeClaim"}, 
        {"name"=>"persistentvolumeclaims/status", "namespaced"=>true, "kind"=>"PersistentVolumeClaim"}, 
        {"name"=>"persistentvolumes", "namespaced"=>false, "kind"=>"PersistentVolume"}, 
        {"name"=>"persistentvolumes/status", "namespaced"=>false, "kind"=>"PersistentVolume"}, 
        {"name"=>"pods", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/attach", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/binding", "namespaced"=>true, "kind"=>"Binding"},
        {"name"=>"pods/exec", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/log", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/portforward", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/proxy", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"pods/status", "namespaced"=>true, "kind"=>"Pod"}, 
        {"name"=>"podtemplates", "namespaced"=>true, "kind"=>"PodTemplate"}, 
        {"name"=>"replicationcontrollers", "namespaced"=>true, "kind"=>"ReplicationController"}, 
        {"name"=>"replicationcontrollers/scale", "namespaced"=>true, "kind"=>"Scale"}, 
        {"name"=>"replicationcontrollers/status", "namespaced"=>true, "kind"=>"ReplicationController"}, 
        {"name"=>"resourcequotas", "namespaced"=>true, "kind"=>"ResourceQuota"}, 
        {"name"=>"resourcequotas/status", "namespaced"=>true, "kind"=>"ResourceQuota"}, 
        {"name"=>"secrets", "namespaced"=>true, "kind"=>"Secret"}, 
        {"name"=>"serviceaccounts", "namespaced"=>true, "kind"=>"ServiceAccount"},
        {"name"=>"services", "namespaced"=>true, "kind"=>"Service"},
        {"name"=>"services/proxy", "namespaced"=>true, "kind"=>"Service"}, 
        {"name"=>"services/status", "namespaced"=>true, "kind"=>"Service"}
      ]
    }
    stub_request(:get, "http://foobar.server/api/v1").to_return(body: discover.to_json)
  end
```

**Comment by grosser** [comment]
... failing with failures from master ... so should be good ...

**Comment by grosser** [comment]
Ready to merge ?

**Comment by simon3z** [comment]
@moolitayer can you review?

File: lib/kubeclient/common.rb [hunk 1]
```diff
@@ -430,6 +422,18 @@ def api
 
     private
 
+    def load_entity_data
```
Comment (moolitayer) [inline lib/kubeclient/common.rb]: perhaps `fetch_entities` would be a better name.

File: lib/kubeclient/common.rb [hunk 2]
```diff
@@ -430,6 +422,18 @@ def api
...
+      JSON.parse(handle_exception { rest_client.get(@headers) })
+    end
+
+    def load_entities
```
Comment (moolitayer) [inline lib/kubeclient/common.rb]: Please move load_entities above load_entity_data

**Comment by moolitayer** [comment]
> Issue 2:
> every new client needs to make a discover call before being useable
> 
> ... I want to be able to share @entities accross all instances or KubeClient by adding some class-level caching ... can do this by overriding load_entities

Note two KubeClient instances doing discovery on different endpoints should have different @entities and methods

File: lib/kubeclient/common.rb (hunk 1 above)
Comment (grosser) [inline lib/kubeclient/common.rb]: done

File: lib/kubeclient/common.rb (hunk 2 above)
Comment (grosser) [inline lib/kubeclient/common.rb]: done

**Comment by moolitayer** [comment]
LGTM :+1: cc @simon3z

**Pull request merged**